### PR TITLE
HPCC-22955 Add caching of public/private key files

### PR DESCRIPTION
**Pull request by RussWhitehead**
Crypto Plugin should cache public/private key files,
since it is costly to load them every time

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@richardkchapman Your concern about a rogue ECL query specifying the PKE file that is already cached seems valid and we should discuss
@ghalliday The documentation for OpenSSL is bad, but in searches I could find anything that indicates these CLoadedKey files cannot be shared. Furthermore my testing did not find any problems.
Please rereview

File: plugins/cryptolib/CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,8 @@ include_directories (
          ${HPCC_SOURCE_DIR}/system/include
          ${HPCC_SOURCE_DIR}/system/jlib
          ${HPCC_SOURCE_DIR}/rtl/include
+         ${HPCC_SOURCE_DIR}/dali/base
```
Comment (ghalliday) [inline plugins/cryptolib/CMakeLists.txt]: These dependencies don't look right.  Why are these now required?

**Comment by ghalliday** [comment]
I'm not sure what benefit is of adding the user to the hash key.  If they passed the same key and passphrase then they will load the key anyway, so I don't think it will provide any extra protection.  On the other hand it is likely to be the same user on calls, so won't make the caching worse.

The main concern would be over whether caching the keys in memory makes it more likely that they could be extracted from the running process.

**Comment by ghalliday** [comment]
@RussWhitehead I forgot to mention you to reassign.

**Comment by RussWhitehead** [comment]
@richardkchapman @ghalliday @jakesmith 
I believe the real problem with this implementation is that, because the key file is opened and read by the execution engine (HPCC User), ANY user that can run a workunit can access any key file accessible to HPCC User. The changes I introduced above will not fix this problem. 
Not sure how to fix this other than removing the ability for an ECL user to specify a filespec, and force them to manage the key files with their own confidential store.
Comments?

**Comment by jakesmith** [comment]
The issue of being able to point to a physical file, and for any user to be able to access it,
yes that's an issue, but not really related to this PR.
That's a problem already (without caching being added).

Either specifying a file path should be prohibited, or it should be validated for authority.
We _could_ turn it into a file::<location> sepc., which can already be tied down via LDAP scope checking.
Or similarly, we could have a separate scheme which allows named key pairs to be referenced and fetched (after passing LDAP checks).

Not sure if WsStore could fit in here somewhere(?) - could it allow groups to share secrets? @rpastrana ?

@RussWhitehead - there should be a separate JIRA covering this issue as not directly to do with this new caching. Can you open a separate JIRA to discuss further?

Re. caching - I think there was concern that it opened up another vulnerability, in that the keys are sitting in memory and in theory an unauthorized user could use them.

Was the main incentive behind the key caching?
Is there 1 area that particularly gains from this time saving?

**Comment by rpastrana** [comment]
Re: WsStore group based restrictions. It wasn't in the design, but it sounds like a reasonable idea. The service would be dependent on the ESP/security manager to provide group info. @RussWhitehead  @jakesmith

**Comment by jakesmith** [comment]
Further discussion is needed in the new JIRA specifically for this issue (of securely getting/sharing secrets/keys), @RussWhitehead - if u can open one.

Re. caching - we can continue to discuss here, @RussWhitehead can u comment on the questions above re. caching.

**Comment by RussWhitehead** [comment]
@jakesmith  Opened https://track.hpccsystems.com/browse/HPCC-23082 to address key file Accessibility

**Comment by RussWhitehead** [comment]
@ghalliday @jakesmith This change restricts the cached key files to the user/key/passphrase that originally referenced it. Any user that subsequently tries to reference the same key file will have to go through the scope check.  Once we add key file scoping, this will make even more sense.

@ghalliday You are correct, having these files cached in memory would make them accessible to snooping. We should discuss the tradeoffs between run time security and performance

File: plugins/cryptolib/CMakeLists.txt (hunk 1 above)
Comment (RussWhitehead) [reply plugins/cryptolib/CMakeLists.txt]: @ghalliday These dependencies were introduced with adding the ICodeContext to the plugin API call

**Comment by ghalliday** [comment]
@RussWhitehead back to you to follow up on the discussion and switch to using thread local variables for the cache.

**Comment by RussWhitehead** [comment]
Replaced by https://github.com/hpcc-systems/HPCC-Platform/pull/13204

**Pull request closed without merging**